internal/utils/gke.go

```diff
@@ -0,0 +1,501 @@
+			return fmt.Errorf(cannotBeNilForNodePoolError, "diskSizeGb", np.Name, config.Name)
+		}
+		if np.Config.DiskType == nil {
+			return fmt.Errorf(cannotBeNilForNodePoolError, "diskType", np.Name, config.Name)
+		}
+		if np.Config.ImageType == nil {
+			return fmt.Errorf(cannotBeNilForNodePoolError, "imageType", np.Name, config.Name)
+		}
+		if np.Config.MachineType == nil {
+			return fmt.Errorf(cannotBeNilForNodePoolError, "machineType", np.Name, config.Name)
+		}
+		if np.Config.Preemptible == nil {
+			return fmt.Errorf(cannotBeNilForNodePoolError, "preemptible", np.Name, config.Name)
+		}
+	}
+
+	return nil
+}
+
+func GetServiceClient(ctx context.Context, credential string) (*gkeapi.Service, error) {
+	ts, err := GetTokenSource(ctx, credential)
+	if err != nil {
+		return nil, err
+	}
+	return GetServiceClientWithTokenSource(ctx, ts)
+}
+
+func GetServiceClientWithTokenSource(ctx context.Context, ts oauth2.TokenSource) (*gkeapi.Service, error) {
+	return gkeapi.NewService(ctx, option.WithHTTPClient(oauth2.NewClient(ctx, ts)))
+}
+
+func GetTokenSource(ctx context.Context, credential string) (oauth2.TokenSource, error) {
+	ts, err := google.CredentialsFromJSON(ctx, []byte(credential), gkeapi.CloudPlatformScope)
+	if err != nil {
+		return nil, err
+	}
+	return ts.TokenSource, nil
+}
+
+func UpdateCluster(config *gkev1.GKEClusterConfig, updateRequest *gkeapi.UpdateClusterRequest) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := GetServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return err
+	}
+
+	_, err = svc.Projects.Locations.Clusters.Update(ClusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName), updateRequest).Context(ctx).Do()
+
+	return err
+}
+
+// LocationRRN returns a Relative Resource Name representing a location. This
+// RRN can either represent a Region or a Zone. It can be used as the parent
+// attribute during cluster creation to create a zonal or regional cluster, or
+// be used to generate more specific RRNs like an RRN representing a cluster.
+//
+// https://cloud.google.com/apis/design/resource_names#relative_resource_name
+func LocationRRN(projectID, location string) string {
+	return fmt.Sprintf("projects/%s/locations/%s", projectID, location)
+}
+
+// ClusterRRN returns an Relative Resource Name of a cluster in the specified
+// region or zone
+func ClusterRRN(projectID, location, clusterName string) string {
+	return fmt.Sprintf("%s/clusters/%s", LocationRRN(projectID, location), clusterName)
+}
+
+// NodePoolRRN returns a Relative Resource Name of a node pool in a cluster in the
+// region or zone for the specified project
+func NodePoolRRN(projectID, location, clusterName, nodePool string) string {
+	return fmt.Sprintf("%s/nodePools/%s", ClusterRRN(projectID, location, clusterName), nodePool)
+}
+
+func BuildUpstreamClusterState(upstreamSpec *gkeapi.Cluster) (*gkev1.GKEClusterConfigSpec, error) {
```

Comment:
I know this is a WIP, but it seems that from `GenerateGkeClusterCreateRequest`, there should be more config pulled from the upstream state.

Reply:
working on that in a separate PR